### PR TITLE
node/bindnode: start of a reflect-based Node implementation

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

**Comment by mvdan** [comment]
To verify that this works:

```
$ cd schema/gen/go
$ go test -tags fluentmirror
PASS
ok      github.com/ipld/go-ipld-prime/schema/gen/go     0.010s
```

File: schema/gen/go/testMaps_test.go
Comment (warpfork) [inline schema/gen/go/testMaps_test.go:62]: Okay, so I realize this was probably a "fasted way to adapt these tests for reuse" thing, but it's a pretty goodsized wat.

I'd like to have these tests not needing to import the bind package.  To do so is just a really weird shape for the import graph.

Comment (warpfork) [reply schema/gen/go/testMaps_test.go:62]: In fact, it would be pretty nice to remove these from the `schema/gen` prefix, since we're now seeing they're definitely more broadly applicable than that (rather than just hopefully so :sweat_smile: ).

I don't know how much work that is / if we want to couple it to this diff.  If we _could_ get these extracted in one PR, and then use it for the new bind features in an immediately subsequent PR, though, that'd be super dreamy.

File: fluent/bind/mirror.go
Comment (warpfork) [inline fluent/bind/mirror.go:1]: I'd move this to `node/bind` instead of `fluent/bind`.

(And possibly declare it as `package bindnode`, to match theme with `package basicnode`... although if you want to argue that `basicnode` shouldn't have been named like that, yeah, okay, maybe so.)

It's a node implementation moreso than a way of working with other node implementations.  So definitely goes under `node/*` rather than `fluent/*`.

Comment (mvdan) [reply fluent/bind/mirror.go:1]: Fair enough. I'll put it under node/bindnode though, because `import ".../bindnode"` is always going to be shorter and simpler than `import bindnode ".../bind"`.

File: schema/gen/go/testMaps_test.go
Comment (mvdan) [reply schema/gen/go/testMaps_test.go:62]: As for the DeepEqual itself, I don't think it's okay for the tests to assume that will work on any ipld.Node implementation. Maybe the "engine" passed to the nodetest package should be an interface, and another method could be "get me the underlying comparable value of this node", or "tell me if these two nodes are equal".

Another option is to declare a semi-standard "Unwrap" optional method on Node types. I have a TODO for that on my code, but I felt like it was too early to bring that up.

File: fluent/bind/mirror.go
Comment (mvdan) [reply fluent/bind/mirror.go]: Done.

File: schema/gen/go/testMaps_test.go
Comment (mvdan) [reply schema/gen/go/testMaps_test.go]: Done; removed all the test changes from this PR, they'll come in a follow-up PR.

**Comment by mvdan** [comment]
I still don't intend to continue writing code/tests/examples until Friday, but if we could review and merge this today, that would help - then tomorrow I can send the tests PR, and I can continue on Friday without being blocked by the stack of changes that are still up for review and might need changes.

File: schema/gen/go/testMaps_test.go
Comment (warpfork) [reply schema/gen/go/testMaps_test.go]: Yes... I think we could also create a generic `Equals` function for `ipld.Node` (traversals of them are well defined after all) and that would be a generally reasonable and useful thing to export even beyond tests.  Or, even more powerfully (and helpful for debugging), a `Diff` feature, though that's more work.

We could also go further and make [`Checker`](https://pkg.go.dev/github.com/frankban/quicktest#Checker) implementations (or, [`Checker`](https://pkg.go.dev/github.com/warpfork/go-wish#Checker), depending on how quickly we migrate testing libraries) which use that feature.

Comment (mvdan) [reply schema/gen/go/testMaps_test.go]: Oh, right, we could have our own `ipld.DeepEqual` which would do the same kind of nested comparison that `reflect.DeepEqual` does, but at the IPLD data model layer instead of Go reflection layer.

Comment (warpfork) [reply schema/gen/go/testMaps_test.go]: Exactly :D

File: node/bindnode/node.go
Comment (warpfork) [inline node/bindnode/node.go:83]: Short notes on what this is for would be helpful.  I see what it does, but harder to see _why_ it's necessary to do the things it does.

I think I see it mostly being used by... code that's related to unions?

Okay, and `PrototypeOnlySchema`.  Which I guess is pretty significant.

When I started writing this comment, I was wondering if the behaviour for unions is being used in a special enough way it should be separated from the rest of this, but I think I've talked myself back out of that already, so this is just a thinking-out-loud note now :)

File: node/bindnode/node.go
Comment (warpfork) [inline node/bindnode/node.go:73]: This is *wild*.  :D  (You already blew my mind a couple days ago when telling me `reflect.StructOf` exists, you probably remember, heh.)

I have some rambling questions as I try to fit my head around when one would use this:

Do we know if this is going to be able to work for all schema typekinds and all their representations?  If there are any typekinds or representation strategies it's *not* going to be able to support, that could be a serious limitation of this entire feature (... possibly enough to make it questionable whether the feature is worth supporting?).

There are going to be some pretty interesting caveats on the usability of this too, right?  Like if it returns a struct... it's going to be some kind of slightly odd anonymous type, which can really only be usefully interacted with via more reflection or our Node interfaces?  ... Okay, I guess that may still sometimes be pretty useful, because it lets one get nice compact native memory layouts without having to use codegen.  That's cool.  But this feature wouldn't really be able to provide a "developer experience" win, right?

**Review by warpfork — COMMENTED** [review]
I'm thinking a hunk of documentation somewhere talking about the pros and cons of different ways to start using this could be important.

There's a huge comparison to be made vs basicnode (e.g. basicnode is general, but not efficient, and has zero DX bonuses), and a huge comparison to be made vs codegen (e.g. codegen is superfast, and has some DX bonuses (but other drawbacks), but has high friction to start using, and increases binary sizes)... and there's also _several_ different ways of using this reflection stuff, if I'm reading right (some of which has certain DX bonsues, and some which doesn't).

File: node/bindnode/node.go
Comment (warpfork) [inline node/bindnode/node.go:174]: So there is basically never going to be a way to avoid this alloc in here, right?  Short of, maybe, some very fancy use of `unsafe`?

Maybe an acceptable cost for this system, just thinking outloud about how docs should talk about performance of this system.

**Comment by mvdan** [comment]
> I'm thinking a hunk of documentation somewhere talking about the pros and cons of different ways to start using this could be important.

For sure. I plan to work on runnable examples after this, then finishing the important TODOs, and lastly high-level docs - once we're happy with the APIs.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:83]: > Okay, and `PrototypeOnlySchema`. Which I guess is pretty significant.

Right, that's the only use case. When supplying both a schema and a Go type, inferGoType should never be used - that API entrypoint is just not properly done yet.

And yeah, unions are special because they need the extra "tag", but otherwise they are handled like any other type.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:73]: > Do we know if this is going to be able to work for all schema typekinds and all their representations?

I'm pretty sure it should support all schema types, and by extension all representations too - representations are just some logic atop the same inferred Go types.

The big caveat is "what if I want my Go type to be slightly different", such as having your own Go type with methods, or using different types like `int` instead of `int64` for a field if you're absolutely sure the values will be small. That's where the answer will be "bring your own Go type and use the other API".

Comment (mvdan) [reply node/bindnode/node.go:73]: Regarding anonymous types: it certainly has some disadvantages, like what I mentioned above. But the advantage is... it just works, without extra effort. No codegen, no having to declare the types yourself. If you have schema X and you want to convert data encoded in dagjson to dagcbor following the schema, you can do all of that with just the schema and PrototypeOnlySchema. The Go value you will hold in between will use anonymous types built at run-time, but for many simple use-cases, that's fine.

Comment (mvdan) [reply node/bindnode/node.go:73]: But yes, for most real use cases like go-codec-dagpb where the type should be easy to use (methods, type switches, better error messages, etc), I do expect the code to provide both the schema and a Go type maintained by a human.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:174]: I mean, the Builder/Assembler themselves already alloc, as they are interfaces. And this is one alloc per built Node, so I don't think it's particularly worrysome.

I am not worrying about performance just yet. My high-level intuition is that I'll be able to allocate about the same, or perhaps a little less, than the codegen. Another question will be the actual performance, where allocs are not the only factor.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:237]: Maybe. For the time being I think the only way to customize this will be... to bring your own Go type, and change the names to whatever you want, as long as the "shape" of the Go type still matches the schema.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:201]: I'm not allowed to spawn an anonymous schema string type, so I made up one :)

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:264]: I'm not worrying about alloc trickery just yet :) When I do, it will be driven by benchmarks anyway, of which I have none so far.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:284]: I'll add a TODO about this, thanks. There are just a lot of edge cases to worry about, so I want to make sure this is covered by a test first.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:355]: I imagined so, but I thought this PR was large enough as is :P 

Also, I wonder what would be the best way to test this, because the schema/gen/go tests don't use it. Maybe some of the selector test cases could also be made "generic" to run over both schema/gen/go and node/nodebind.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:215]: Because I initially called all these nodeWrapper, assemblerWrapper, etc. The names just got veeeery long and I didn't like them, so I went for _node, _assembler, etc.

I'm not too fond of `n _node` because `n` is usually `n int`, and `p _prototype` is usually `p []byte`, and `s _structAssembler` is usually `s string`... so I was a bit lazy and left that out for now, defaulting to `w` for all of them, which is almost never used in a standard way, and still gives the hint that these are "wrappers".

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:309]: This would work if we assume we always infer the type, but longer term I want to allow the user more control. You can imagine allowing the user to provide a custom type to back a union by providing methods like `Get() (tag int, value interface{})` and `Put(tag int, value interface{})`.

File: node/bindnode/node.go
Comment (warpfork) [reply node/bindnode/node.go:359]: Yep, those test batteries should still grow.  IIRC I did some work to make sure both iterators and lookup methods got exercised for the same data (with the newest form of tests, anyway), but maybe there's still a bit more to do there to make sure _all_ the lookup method variants are exercised too.

Getting selector tests to run over arbitrary nodes would also be super good.  I think there's a small barrier there right now, in that those tests only take one NodePrototype at the moment, and assume they can use it for both selector data and the test data, and that needs fixing... tractable but definitely enough of a diff to make it its own little quest.

File: node/bindnode/node.go
Comment (warpfork) [inline node/bindnode/node.go:672]: I think I would use node.Length here, yeah.

Also: I've been considering extracting this whole hunk below to a generic function like, say, `ipld.BlindCopyMap`, because it does come up in a couple places now.  (The other big example is similarly, in the AssignNode methods emitted by codegen, this logic exists as a fallback path right after trying to do direct copy for known types.)  So if that looks reasonable to extract, go for it.

File: node/bindnode/node.go
Comment (warpfork) [inline node/bindnode/node.go:749]: I think this TODO is probably overoptimistic.  If a field has either the nullable or optional properties in the schema, checking the go field for nil doesn't provide any useful information about if it was assigned or not.

Keeping a bitfield explicitly is also just going to be a lot simpler code, I suspect.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:672]: I sorta left these TODOs as reminders for future optimizations, and as said before, I have no benchmarks. From what you say, we might dedup this anyway :)

I'll add this refactor as another TODO.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:749]: Ah, you're right, I keep forgetting about AssignNull which does assign into a node, but explicitly keeps it null.

I still think there's room for improvement here potentially, but I also don't think it's going to be a bottleneck soon. It would always be possible to avoid an extra allocation for most structs by backing this slice with a `[16]bool` in the struct., for example.

I'll leave this TODO about reminding myself to look at this at some point.

**Comment by mvdan** [comment]
I added a runnable example, but I don't get why it's failing...

```
--- FAIL: ExamplePrototypeOnlySchema (0.00s)
got:
{
        "Name": "Michael",
        "Friends": [
                "Sarah",
                "Alex"
        ]
}
want:
{
        "Name": "Michael",
        "Friends": [
                "Sarah",
                "Alex"
        ]
}
FAIL
exit status 1
FAIL    github.com/ipld/go-ipld-prime/node/bindnode     0.002s
```

File: node/bindnode/repr.go
Comment (warpfork) [inline node/bindnode/repr.go:184]: Should be `AppropriateKind: ipld.KindSet_Recursive` among other things.  Probably this whole method is actually a todo right now?

File: node/bindnode/node.go
Comment (willscott) [inline node/bindnode/node.go:314]: more specificity on attempted vs reality would indeed be useful here :)

File: node/bindnode/node.go
Comment (willscott) [reply node/bindnode/node.go:362]: I'd be in favor to see this done in a subsequent PR once there are tests for that path.

File: node/bindnode/node.go
Comment (willscott) [inline node/bindnode/node.go:380]: this one probably does need to be in this PR

File: node/bindnode/node.go
Comment (willscott) [inline node/bindnode/node.go:1112]: the current Unwrap seems like a reasonable interface to me.

File: node/bindnode/node.go
Comment (willscott) [inline node/bindnode/node.go:1104]: given the length of this file, it might be worth splitting it into per-type files. e.g `node_union`, `node_map`, `node_list` for the type-specific assemblers / iterators.

File: node/bindnode/repr.go
Comment (willscott) [inline node/bindnode/repr.go:67]: a comment explaining the need for inbound vs outbound here would be useful.

File: node/bindnode/repr.go
Comment (willscott) [inline node/bindnode/repr.go:153]: this should work when the key is a string node, right?

File: node/bindnode/repr.go
Comment (mvdan) [reply node/bindnode/repr.go:184]: Yeah, a lot of these are copy pasted and TODOs. I'll make it explicit.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:69]: The intent is to make things explicit. Prototype vs PrototypeSchema could also be relatively ambiguous, if one doesn't look at the signatures.

I don't feel strongly about these yet and most of the exposed API is still TODO, so I'm happy to take ideas. I'll probably focus on more tests and finishing the code first before worrying about names and docs, though :)

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:25]: Ah, good catch. The original API was like `fluent.WrapNoSchema` etc.

Comment (mvdan) [reply node/bindnode/node.go:25]: Ah, wait, we do still have WrapNoSchema. I just haven't done anything with it yet, because the codegen tests only require prototypes.

But this doc still needs updating.

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:380]: It's hard to draw the line. This PR is huge as is, so I drew the line at "what is necessary to make the tests pass". I plan to continue fixing TODOs as I add more tests in future PRs. If we decide that this many TODOs are not acceptable in master, then I'd pretty easily slip into 5k-line PR territory :)

File: node/bindnode/node.go
Comment (mvdan) [reply node/bindnode/node.go:1104]: Splitting is a good point. I just kept going and didn't really think about file sizes that much.

Maybe node_scalar and node_recursive would be better for now, and later on I can split recursive into map/list/union if necessary. It feels a bit silly to have e.g. node_bool or node_int.

File: node/bindnode/repr.go
Comment (mvdan) [reply node/bindnode/repr.go:67]: These are pretty much just hacks, since the schema API gives the mappings one way but not the other. I'll add a TODO about working this out with Eric - I didn't add docs because I don't intend to maintain nor expose this code here.

File: node/bindnode/repr.go
Comment (mvdan) [reply node/bindnode/repr.go:153]: Indeed, but the same I mentioned elsewhere about tests and TODOs applies :) This PR is by no means attempting to be complete

**Comment by mvdan** [comment]
@warpfork are you up for merging today? The example is passing, and I've replied to all the review comments. I know I replied to many with "yep, in the next PRs", but I really do want to work incrementally rather than keep expanding this PR.

**Review by warpfork — APPROVED** [review]
Yep, I'm good with this merging.  There's work still to do, but it's all encapsulated in one package, so we should not have difficulty making it reasonably clear in the interim.

**Pull request merged**
